File: chap09/run-model9-3.R
library(cmdstanr)
library(dplyr)

set.seed(123)
d <- read.csv(file='input/data-tennis-2017.csv')
N <- max(d)
G <- nrow(d)
data <- list(N=N, G=G, g2L=d$Loser, g2W=d$Winner)

model <- cmdstan_model(stan_file='model/model9-3.stan')
fit <- model$sample(data=data, seed=123, 
                    init=function(){ list(mu=rnorm(N, 0, 0.05)) },
                    parallel_chains=4)
fit$save_object(file='output/result-model9-3.RDS')

mu_ms <- fit$draws('mu', format='matrix')
qua <- apply(mu_ms, 2, quantile, prob=c(0.05, 0.5, 0.95))
d_est <- data.frame(NID=1:N, t(qua), check.names=FALSE)
d_top5 <- d_est %>% slice_max(`50%`, n=5)
#         NID        5%      50%      95%
# mu[54]   54 1.6515475 2.142635 2.701341
# mu[117] 117 1.4823785 1.891510 2.354333
# mu[39]   39 0.8753171 1.357045 1.863492
# mu[174] 174 0.9058864 1.256475 1.674837
# mu[35]   35 0.7998884 1.238130 1.713748

s_pf_ms <- fit$draws('s_pf', format='matrix')
qua <- apply(s_pf_ms, 2, quantile, prob=c(0.05, 0.5, 0.95))
d_est <- data.frame(NID=1:N, t(qua), check.names=FALSE)
d_top3 <- d_est %>% slice_max(`50%`, n=3)
d_bot3 <- d_est %>% slice_min(`50%`, n=3)
#           NID        5%      50%      95%
# s_pf[94]   94 0.7668348 1.248740 1.872305
# s_pf[147] 147 0.7287111 1.225655 1.863775
# s_pf[81]   81 0.7037272 1.187765 1.821020

#           NID        5%      50%      95%
# s_pf[8]     8 0.4114912 0.700248 1.124152
# s_pf[117] 117 0.4165578 0.724443 1.163592
# s_pf[24]   24 0.4573924 0.790498 1.289026

File: chap09/run-model9-3.py
import pandas
import numpy as np
import cmdstanpy

np.random.seed(123)
d = pandas.read_csv('input/data-tennis-2017.csv')
N = np.max(d.to_numpy())
G = len(d)
data = {'N':N, 'G':G, 'g2L':d.Loser, 'g2W':d.Winner}

def init_fun():
    return {'mu':np.random.normal(loc=0, scale=0.05, size=N)}

model = cmdstanpy.CmdStanModel(stan_file='model/model9-3.stan')
fit = model.sample(data=data, seed=123, parallel_chains=4, inits=init_fun())
fit.save_csvfiles('output/result-model9-3')

mu_ms = fit.stan_variable(var='mu')
qua = np.quantile(mu_ms, [0.05, 0.50, 0.95], axis=0)
d_est = pandas.DataFrame(np.column_stack([np.arange(1,N+1), qua.T]), \
    columns=['NID', '10%', '50%', '90%'])
d_top5 = d_est.nlargest(5, '50%')
#        NID       10%       50%       90%
# 53    54.0  1.647616  2.145335  2.758852
# 116  117.0  1.495580  1.890695  2.370836
# 38    39.0  0.864339  1.350435  1.882290
# 173  174.0  0.894896  1.260735  1.645432
# 34    35.0  0.811560  1.235450  1.688341

s_pf_ms = fit.stan_variable(var='s_pf')
qua = np.quantile(s_pf_ms, [0.05, 0.50, 0.95], axis=0)
d_est = pandas.DataFrame(np.column_stack([np.arange(1,N+1), qua.T]), \
    columns=['NID', '10%', '50%', '90%'])
d_top3 = d_est.nlargest(3, '50%')
d_bot3 = d_est.nsmallest(3, '50%')
#        NID       10%       50%       90%
# 93    94.0  0.745436  1.246845  1.872001
# 146  147.0  0.728718  1.212850  1.853632
# 80    81.0  0.714567  1.192350  1.819750

#        NID       10%       50%       90%
# 7      8.0  0.403085  0.705031  1.138887
# 116  117.0  0.418390  0.726276  1.158035
# 23    24.0  0.447273  0.788228  1.293023
